File: trees_and_graphs/all_nodes_distance_k_in_binary_tree.java
class Solution
{
    Map<TreeNode, TreeNode> parents = new HashMap<>();

    public void dfs(TreeNode node, TreeNode parent)
    {
        if (node == null)
        {
            return;
        }

        parents.put(node, parent);
        dfs(node.left, node);
        dfs(node.right, node);
    }

    public List<Integer> distanceK(TreeNode root, TreeNode target, int k)
    {
        dfs(root, null);

        Queue<TreeNode> queue = new LinkedList<>();
        Set<TreeNode> seen = new HashSet<>();
        queue.add(target);
        seen.add(target);

        int distance = 0;
        while (!queue.isEmpty() && distance < k)
        {
            int currentLength = queue.size();
            for (int i = 0; i < currentLength; i++)
            {
                TreeNode node = queue.remove();
                for (TreeNode neighbor : new TreeNode[]{
                    node.left, node.right, parents.get(node)})
                {
                    if (neighbor != null && !seen.contains(neighbor))
                    {
                        seen.add(neighbor);
                        queue.add(neighbor);
                    }
                }
            }

            distance++;
        }

        List<Integer> ans = new ArrayList<>();
        for (TreeNode node : queue)
        {
            ans.add(node.val);
        }

        return ans;
    }
}

File: trees_and_graphs/all_nodes_distance_k_in_binary_tree.py
import treenode

from collections import deque

def distanceK(
    root: treenode.TreeNode, target: treenode.TreeNode, k: int
)-> list[int]:
    def dfs(node, parent):
        if not node:
            return
        
        node.parent = parent
        dfs(node.left, node)
        dfs(node.right, node)
    
    dfs(root, None)

    queue = deque([target])
    seen = {target}
    distance = 0

    while queue and distance < k:
        current_length = len(queue)

        for _ in range(current_length):
            node = queue.popleft()

            for neighbor in [node.left, node.right, node.parent]:
                if neighbor and neighbor not in seen:
                    seen.add(neighbor)
                    queue.append(neighbor)
        
        distance += 1
    
    return [node.val for node in queue]